src/tiled/toolspecifictoolbar.cpp

```diff
@@ -0,0 +1,162 @@
+
+#include "abstracttool.h"
+#include "bucketfilltool.h"
+#include "mapeditor.h"
+#include "stampbrush.h"
+
+#include <QEvent>
+#include <QToolButton>
+
+namespace Tiled {
+namespace Internal {
+
+ToolSpecificToolBar::ToolSpecificToolBar(QWidget *parent, MapEditor *mapEditor)
+    : QToolBar(parent)
+    , mMapEditor(mapEditor)
+    , mDiceIcon(QLatin1String(":images/24x24/dice.png"))
+    , mFlipHorizontalIcon(QLatin1String(":images/24x24/flip-horizontal.png"))
+    , mFlipVerticalIcon(QLatin1String(":images/24x24/flip-vertical.png"))
+    , mRotateLeftIcon(QLatin1String(":images/24x24/rotate-left.png"))
+    , mRotateRightIcon(QLatin1String(":images/24x24/rotate-right.png"))
+{
+    setObjectName(QLatin1String("ToolSpecificToolBar"));
+    setWindowTitle(tr("Tool Specific Tool Bar"));
+    setToolButtonStyle(Qt::ToolButtonFollowStyle);
+
+    mDiceIcon.addFile(QLatin1String(":images/32x32/dice.png"));
+    mFlipHorizontalIcon.addFile(QLatin1String(":images/32x32/flip-horizontal.png"));
+    mFlipVerticalIcon.addFile(QLatin1String(":images/32x32/flip-vertical.png"));
+    mRotateLeftIcon.addFile(QLatin1String(":images/32x32/rotate-left.png"));
+    mRotateRightIcon.addFile(QLatin1String(":images/32x32/rotate-right.png"));
+
+    retranslateUi();
+}
+
+void ToolSpecificToolBar::changeEvent(QEvent *event)
+{
+    QToolBar::changeEvent(event);
+    switch (event->type()) {
+    case QEvent::LanguageChange:
+        retranslateUi();
+        break;
+    default:
+        break;
+    }
+}
+
+void ToolSpecificToolBar::setSelectedTool(AbstractTool *tool)
+{
+    qDeleteAll(mButtons);
+    mButtons.clear();
+
+    clear();
+
+    if (!tool)
+        return;
+
+    StampBrush *stampBrush = qobject_cast<StampBrush*>(tool);
+    if (stampBrush) {
+        addRandomTool(stampBrush->random());
+        addFlippingTools();
+        addRotatingTools();
+    }
+
+    BucketFillTool *bucketFillTool = qobject_cast<BucketFillTool*>(tool);
+    if (bucketFillTool) {
+        addRandomTool(bucketFillTool->random());
+        addFlippingTools();
+        addRotatingTools();
+    }
+}
+
+void ToolSpecificToolBar::onOrientationChanged(Qt::Orientation orientation)
+{
+    setToolButtonStyle(orientation == Qt::Horizontal ? Qt::ToolButtonFollowStyle :
+                                                       Qt::ToolButtonIconOnly);
+}
+
+void ToolSpecificToolBar::addRandomTool(bool checked)
+{
+    QToolButton *mRandomButton = new QToolButton(this);
+    mRandomButton->setIcon(mDiceIcon);
+    mRandomButton->setCheckable(true);
+    mRandomButton->setToolTip(tr("Random Mode"));
+    mRandomButton->setShortcut(QKeySequence(tr("D")));
+    mRandomButton->setChecked(checked);
+
+    addWidget(mRandomButton);
+    mButtons.append(mRandomButton);
+
+    connect(mRandomButton, &QToolButton::toggled, mMapEditor, &MapEditor::random);
+}
+
+void ToolSpecificToolBar::addFlippingTools()
+{
+    QToolButton *mFlipHorizontalButton = new QToolButton(this);
```

Comment:
Try to rely on `QAction` when possible (and I think it's possible for all actions you're currently adding). That way, you can instantiate the action once and just add it to the tool bar when applicable. It won't get deleted upon calling `clear`, which is good.